pkg/operator2/route.go

```diff
@@ -178,3 +179,74 @@ func isIngressAdmitted(ingress routev1.RouteIngress) bool {
 func ingressToHost(ingress *configv1.Ingress) string {
 	return targetName + "." + ingress.Spec.Domain
 }
+
+func validateRouterSecret(routerSecret *corev1.Secret, ingress *configv1.Ingress) (string, error) {
+	// be careful not to print the routerSecret even when it is empty
+
+	if len(routerSecret.Data) == 0 {
+		return "EmptyRouterSecret", fmt.Errorf("router secret is empty")
+	}
+
+	domain := ingress.Spec.Domain
+	if len(domain) == 0 {
+		return "NoIngressDomain", fmt.Errorf("ingress config has no domain")
+	}
+
+	pemCerts := routerSecret.Data[domain]
+	if len(pemCerts) == 0 {
+		return "NoIngressDataRouterSecret", fmt.Errorf("router secret has no data for ingress domain %s", domain)
+	}
+
+	certificates, err := crypto.CertsFromPEM(pemCerts)
+	if err != nil {
+		return "InvalidPEMRouterSecret", fmt.Errorf("router secret contains invalid PEM data: %v", err)
+	}
+
+	opts := x509.VerifyOptions{
+		DNSName:       ingressToHost(ingress),
+		Intermediates: x509.NewCertPool(),
+		Roots:         x509.NewCertPool(),
+	}
+
+	var hasCA, hasCert bool
+
+	for _, certificate := range certificates {
+		if !certificate.IsCA {
+			continue
+		}
+
+		// consider self-signed CAs as roots
+		if bytes.Equal(certificate.RawIssuer, certificate.RawSubject) {
+			opts.Roots.AddCert(certificate)
+			hasCA = true
+			continue
+		}
+
+		// consider all other CAs as intermediates
+		opts.Intermediates.AddCert(certificate)
+	}
+
+	for _, certificate := range certificates {
+		if certificate.IsCA {
+			continue
+		}
+
+		if _, err := certificate.Verify(opts); err != nil {
+			klog.V(4).Infof("cert %s failed verification: %v", certificate.Subject.String(), err)
+			continue
+		}
+
+		hasCert = true
+		break
+	}
+
+	if !hasCA {
+		return "NoCARouterSecret", fmt.Errorf("router secret contains no CA")
```

Comment:
"NoRootCARouterSecret", "contains no root CA" -> it might be containing CAs, just not roots

Wouldn't intermediates be enough for our use-case?